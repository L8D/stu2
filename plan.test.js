const Promise = require('bluebird')
const moment = require('moment-timezone')
const flatten = require('lodash.flatten')

const {
  parseTaskDeadlines,
  process
} = require('./processor')

const exampleTasks = [
  {
    deadline: '7am',
    objective: 'wake up'
  },

  {
    deadline: '7:30am',
    objective: 'take theanine'
  },

  {
    deadline: '8am',
    objective: 'shower'
  },

  {
    deadline: '9am',
    objective: 'meditate'
  },

  {
    deadline: '12pm',
    objective: 'get out'
  },

  {
    deadline: '1pm',
    objective: 'eat lunch'
  },

  {
    deadline: '1:05pm',
    objective: 'log lunch'
  },

  {
    deadline: '6pm',
    objective: 'form a plan for getting home and eating dinner'
  },

  {
    deadline: '8pm',
    objective: 'go home and eat dinner'
  },

  {
    deadline: '8:20pm',
    objective: 'write'
  },

  {
    deadline: '8:40pm',
    objective: 'draw'
  },

  {
    deadline: '9pm',
    objective: 'wash'
  },

  {
    deadline: '9:30pm',
    objective: 'meditate'
  }
]

describe('executor', () => {
  it('should advance the state given the example plan', async () => {
    const timeFormats = ['h:mma', 'ha']
    const date = '2018-11-23'
    const timezone = 'America/Chicago'

    // every five minutes
    const heartbeats = []
    let time = moment.tz(`${date}T06:00:00`, timezone)

    for (let i = 0; i < 300; i++) {
      heartbeats.push({
        type: 'heartbeat',
        createdAt: time.toISOString(),
        timezone
      })

      time = time.add(5, 'minutes')
    }

    const tasks = parseTaskDeadlines(exampleTasks, {
      date,
      timezone
    })

    const db = {
      cache: {},
      getAsync: (key) => Promise.resolve(db.cache[key])
    }

    const results = []

    for (const message of heartbeats) {
      for (const action of await process(tasks, message, db)) {
        results.push(action)

        // statefully handle 'set' operations between heartbeats
        if (action.type === 'set') {
          db.cache[action.key] = action.value
        }
      }
    }

    expect(results).toMatchSnapshot()
  })
})
