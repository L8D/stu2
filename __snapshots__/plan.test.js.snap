// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executor should advance the state given the example plan 1`] = `
Array [
  Object {
    "deadline": "6:15",
    "sentAt": "2018-11-23T06:00:00-06:00",
    "type": "outbound",
    "value": "You need to take theanine. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:1:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "6:30am",
    "sentAt": "2018-11-23T06:15:00-06:00",
    "type": "outbound",
    "value": "You need to wash. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:2:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "7:45am",
    "sentAt": "2018-11-23T06:30:00-06:00",
    "type": "outbound",
    "value": "You need to meditate. You have 75 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:3:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8am",
    "sentAt": "2018-11-23T07:45:00-06:00",
    "type": "outbound",
    "value": "You need to hack on sudomesh. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:4:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:15am",
    "sentAt": "2018-11-23T08:00:00-06:00",
    "type": "outbound",
    "value": "You need to hack on dat; hyperledger fabric. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:5:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:30am",
    "sentAt": "2018-11-23T08:15:00-06:00",
    "type": "outbound",
    "value": "You need to get ready leave. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:6:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:55am",
    "sentAt": "2018-11-23T08:30:00-06:00",
    "type": "outbound",
    "value": "You need to catch the 8:55am C train towards 168 St. You have 25 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:7:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "10:15am",
    "sentAt": "2018-11-23T08:55:00-06:00",
    "type": "outbound",
    "value": "You need to hack on infrastructure. You have 80 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:8:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "11am",
    "sentAt": "2018-11-23T10:15:00-06:00",
    "type": "outbound",
    "value": "You need to hack on ghostwriter. You have 45 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:9:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "12pm",
    "sentAt": "2018-11-23T11:00:00-06:00",
    "type": "outbound",
    "value": "You need to plan lunch. You have 60 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:10:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "1pm",
    "sentAt": "2018-11-23T12:00:00-06:00",
    "type": "outbound",
    "value": "You need to eat lunch. You have 60 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:11:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "1:05pm",
    "sentAt": "2018-11-23T13:00:00-06:00",
    "type": "outbound",
    "value": "You need to log lunch. You have 5 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:12:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "1:45pm",
    "sentAt": "2018-11-23T13:05:00-06:00",
    "type": "outbound",
    "value": "You need to hack on stu. You have 40 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:13:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "2:45pm",
    "sentAt": "2018-11-23T13:45:00-06:00",
    "type": "outbound",
    "value": "You need to participate in PLP. You have 60 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:14:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "5:00pm",
    "sentAt": "2018-11-23T14:45:00-06:00",
    "type": "outbound",
    "value": "You need to hack on anki. You have 135 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:15:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "5:15pm",
    "sentAt": "2018-11-23T17:00:00-06:00",
    "type": "outbound",
    "value": "You need to form a plan for eating dinner. You have 15 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:16:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "6:30pm",
    "sentAt": "2018-11-23T17:15:00-06:00",
    "type": "outbound",
    "value": "You need to eat dinner. You have 75 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:17:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "7:30pm",
    "sentAt": "2018-11-23T18:30:00-06:00",
    "type": "outbound",
    "value": "You need to go home. You have 60 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:18:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:00pm",
    "sentAt": "2018-11-23T19:30:00-06:00",
    "type": "outbound",
    "value": "You need to write. You have 30 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:19:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:20pm",
    "sentAt": "2018-11-23T20:00:00-06:00",
    "type": "outbound",
    "value": "You need to draw. You have 20 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:20:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "8:40pm",
    "sentAt": "2018-11-23T20:20:00-06:00",
    "type": "outbound",
    "value": "You need to wash. You have 20 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:21:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "9:00pm",
    "sentAt": "2018-11-23T20:40:00-06:00",
    "type": "outbound",
    "value": "You need to meditate. You have 20 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:22:announcement",
    "type": "set",
    "value": true,
  },
  Object {
    "deadline": "9:30pm",
    "sentAt": "2018-11-23T21:00:00-06:00",
    "type": "outbound",
    "value": "You need to read a book. You have 30 minutes.",
  },
  Object {
    "key": "2018-11-23:tasks:23:announcement",
    "type": "set",
    "value": true,
  },
]
`;
